### PR TITLE
[W5.4][W11-A2]Zhan Yu

### PR DESCRIPTION
**Pull request by ZY-Ang**
Added sort functionality, (+help,tests,docs) as per my level1 implementation. Users can sort using the list command or find command.

### CODE REVIEW

File: src/seedu/addressbook/commands/SortableCommand.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            for (String sortArgument : sortArguments) {
+                if (c == 0) {
+                    c = sortArgumentCompareValue(sortArgument, person1, person2);
+                }
+            }
+            return c;
+        });
+
+        return sortedPersons;
+    }
```
Comment (JYL123) [inline src/seedu/addressbook/commands/SortableCommand.java]: I like the implementation of this comparator.

File: src/seedu/addressbook/commands/SortableCommand.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            case Name.PREFIX:
+                return person1.getName().toString().compareTo(person2.getName().toString());
+                
+            case Phone.PREFIX:
+                return person1.getPhone().toString().compareTo(person2.getPhone().toString());
+                
+            case Email.PREFIX:
+                return person1.getEmail().toString().compareTo(person2.getEmail().toString());
+                
+            case Address.PREFIX:
```
Comment (JYL123) [inline src/seedu/addressbook/commands/SortableCommand.java]: I am a bit confused, but is it necessary to compare prefix for sorting ?

File: test/java/seedu/addressbook/data/AddressBookTest.java [hunk 3]
```diff
@@ -189,4 +191,14 @@ private boolean isTagObjectInAddressBookList(Tag tagToCheck, AddressBook address
...
+    
+    @Test
+    public void testPreSortingConstants() {
+        assertEquals(Address.PREFIX, "a/");
+        assertEquals(Phone.PREFIX, "p/");
+        assertEquals(Email.PREFIX, "e/");
+        assertEquals(Name.PREFIX, "n/");
+        assertEquals(SortableCommand.SORT_ASCENDING_POSTFIX_WORD, "asc");
+        assertEquals(SortableCommand.SORT_DESCENDING_POSTFIX_WORD, "desc");
+    }
```
Comment (JYL123) [inline test/java/seedu/addressbook/data/AddressBookTest.java]: Cool, sorting can be both "asc" and "desc"

File: src/seedu/addressbook/commands/SortableCommand.java (hunk 2 above)
Comment (ZY-Ang) [reply src/seedu/addressbook/commands/SortableCommand.java]: Yep, for example, if I detected a Phone.PREFIX in the argument, I want to use the firstPerson's phone and compare it to the secondPerson's phone. Then I compare the next prefix I found. That way I will get a list sorted by x then by y then by z then by ...

Comment (JYL123) [reply src/seedu/addressbook/commands/SortableCommand.java]: I see. Thanks for explaining.

File: test/java/seedu/addressbook/commands/SortableCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+
+public class SortableCommandTest {
+
+    private SortableCommand sortableCommand;
+    
+    @Before
+    public void setUp() { sortableCommand = new SortableCommand(null); }
```
Comment (okkhoy) [inline test/java/seedu/addressbook/commands/SortableCommandTest.java]: please take a look at the coding guidelines for the spacing

**Comment by okkhoy** [comment]
@ZY-Ang 
Some comments added. Please ack & close the PR after reading comments.

File: src/seedu/addressbook/commands/SortableCommand.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
+package seedu.addressbook.commands;
+
+
+import seedu.addressbook.data.person.Name;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.person.Email;
+import seedu.addressbook.data.person.Address;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (ZY-Ang) [reply src/seedu/addressbook/commands/SortableCommand.java]: crap

**Comment by ZY-Ang** [comment]
@okkhoy Alr updated help. Just needed to change the message usage. Should be the changes in SORT_USAGE. Didn't need to actually update the help implementation as the sorting is done from within list/find

**Pull request closed without merging**